src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala

```diff
@@ -93,23 +103,118 @@ object PrecompiledContracts {
+      15 + 3 * wordsForBytes(inputData.size)
+  }
+
+  object ModExp extends PrecompiledContract {
+
+    private val lengthBytes = 32
+    private val totalLengthBytes = 3 * lengthBytes
+    private val GQUADDIVISOR = 20
+
+    def exec(inputData: ByteString): ByteString = {
+      val baseLength = getLength(inputData, 0)
+      val expLength = getLength(inputData, 1)
+      val modLength = getLength(inputData, 2)
+
+      if (baseLength == 0 && modLength == 0)
+        ByteString.empty
+      else {
+        val base = getNumber(inputData, totalLengthBytes, baseLength)
+        val exp = getNumber(inputData, safeAdd(totalLengthBytes, baseLength), expLength)
+        val mod = getNumber(inputData, safeAdd(totalLengthBytes, safeAdd(baseLength, expLength)), modLength)
+
+        if (mod == 0) {
+          ByteString.empty
+        } else {
+          val result = base.modPow(exp, mod)
+          ByteString(ByteUtils.bigIntegerToBytes(result.bigInteger, modLength))
+        }
+      }
+    }
+
+    def gas(inputData: ByteString): BigInt = {
+      val baseLength = getLength(inputData, 0)
+      val expLength = getLength(inputData, 1)
+      val modLength = getLength(inputData, 2)
+
+      val expBytes =
+        inputData.slice(
+          safeAdd(totalLengthBytes, baseLength),
+          safeAdd(safeAdd(totalLengthBytes, baseLength), expLength))
+
+      val multComplexity = getMultComplexity(math.max(baseLength, modLength))
+
+      val adjExpLen = adjExpLength(expBytes, expLength)
+
+      val gasCost = multComplexity * math.max(adjExpLen , 1) / GQUADDIVISOR
+
+      gasCost
+    }
+
+    private def adjExpLength(expBytes: ByteString, expLength: Int): Long = {
+      val expHead =
+        if (expLength <= lengthBytes)
+          expBytes.padTo(expLength, 0.toByte)
+        else
+          expBytes.take(lengthBytes).padTo(lengthBytes, 0.toByte)
+
+
+      val msb = math.max(ByteUtils.toBigInt(expHead).bitLength - 1, 0)
+
+      if (expLength <= lengthBytes) {
+          msb
+      } else {
+          8L * (expLength - lengthBytes) + msb
+      }
+    }
+
+    private def getLength(bytes: ByteString, position: Int): Int = {
+      val start = position * lengthBytes
+      safeInt(ByteUtils.toBigInt(bytes.slice(start, start + lengthBytes)))
+    }
+
+    private def getNumber(bytes: ByteString, offset: Int, length: Int): BigInt = {
+      val number = bytes.slice(offset, safeAdd(offset, length)).padTo(length, 0.toByte)
+      ByteUtils.toBigInt(number)
+    }
+
+    private def safeInt(value: BigInt): Int =
+      if (value.isValidInt)
+        value.toInt
+      else
+        Integer.MAX_VALUE
+
+    private def safeAdd(a: Int, b: Int): Int = {
+      safeInt(BigInt(a) + BigInt(b))
+    }
+
+    private def getMultComplexity(x: BigInt): BigInt = {
+      val x2 = x * x
+
+      if (x <= 64)
+        x2
+      else if (x <= 1024)
+        x2 / 4 + 96 * x - 3072
+      else
+        x2 / 16 + 480 * x - 199680
```

Comment:
maybe leave a comment why this specific numbers are used

Reply:
hmm i will leave a link to the spec on top as all is defined there